Mixins - Add mgd-php@2. Use tighter search to ignore nested submodules.

totten opened the pull request:
Overview
----------------------------------------

The mixin `mgd-php@1` helps extensions autoload `*.mgd.php` files. The original policy is quite open, allowing these files to be placed anywhere. However, if you have overlapping extensions (e.g. nested extensions or submodules), then this can become actively harmful: the `*.mgd.php` from a child extension can be loaded by its parent extension.

This defines a new iteration (mixin `mgd-php@2`) with a tighter search radius.

Before
----------------------------------------

`mgd-php@1` searches the entire subtree underneath the extension.

After
----------------------------------------

`mgd-php@2` searches specific folders which are commonly used for `*.mgd.php` files:

* __Root Files__ Files in the top-level folder of the extension (`*.mgd.php`)
* __In Situ Files__: Files that are adjacent to relevant code (e.g. `api/**.mgd.php`, `CRM/**.mgd.php`)
* __Managed Folder__: Files in a general managed folder (`managed/**.mgd.php`)

This has a few effects:

* It skips many folders (e.g. `templates/`, `vendor/`, `settings/`, `schema/`) that are commonly included in extensions but which don't have MGDs . Many of these folders are small (e.g. `settings/`), but some can get large (e.g. `vendor/`).
* It skips unknown/unrecognized folders. In observed-practice and in common-sense, the nested-extensions tend to have bespoke folders. You won't find them under `PARENT_EXT/managed/CHILD_EXT/` or `PARENT_EXT/api/v3/CHILD_EXT/`.

Comments
----------------------------------------

You can look at this topic from a few angles:

1. _False-matches_: The impetus for the PR is to fix a problem where some extensions are vulnerable to false-matches, which cause one extension (e.g. `ext/parent/`) to incorrectly load MGDs from another (e.g. `ext/parent/child/`).
2. _Performance_: If you're changing the search, then you can also reduce the search-space - which is good for performance.
3. _Aesthetics_: When writing an extension, you want to put files in a place that feels sensible.
4. _Deployment_: How much work is required to deploy the update?

The impetus for this PR is to solve the false-match problem. But you can imagine different patches which create different trade-offs. For example:

* Scan all folders but _exclude_ nested extensions. If the search encounters a `$subfolder/` with its own `$subfolder/info.xml`, then skip it. _This requires a longer implementation, and there is little prospect for improving performance._
* Scan `./managed/` exclusively. The implementation is fairly simple, and there's more performance benefit. _But this takes more work to deploy (re:conversions), and it prematurely judges the aesthetic situation (re:sensible organization)._

Originally, when drafting this PR, I started with that last notion -- to search only `./managed/`. But then I reviewed `universe` to gauge actual impacts. 

* https://gist.github.com/totten/a63069fbe262a711a54308e7f21b8bec (*This excludes ext's before 2020. The vast majority of ext's are _recent_, i.e. updated in 2024/2025.*)

This led me to the middle-ground (*this PR*). Compare:

* __Deployment__: 
    * If one *only* accepts `./managed/`, then it would require reorganizing 127/210 exts (60%) -- enough that `civix` would need to automatically convert them. That means more detailed testing (and lower tolerance for mistakes).
    * This middle-ground has very high drop-in compatibility. It only requires reorganizing 14/210 exts (6%).  With this smaller number, `civix` doesn't need to do it automatically. (We could just print a notice and leave the oddballs on `mgd-php@1` until the author manually reorganizes.)
* __Aesthetics__:
    * If we had a plan for ___One Good Layout___, then I would be fine with sorting-out migration via `civix` for  60%+ ext's. But what we actually have are ___Multiple Middling Layouts___ (with prevalence of 10%/48%/49%).
    * The layout of `managed/` is centralized, and that's nice with 2-5 MGDs. It works with any kind of entity. But when you get 10+ MGDs, then it's messy.
    * The in-situ layout scales (*more structured*), but it only works with PHP classes and APIv3 files. There is no comparable structure for saved-searches, option-groups, etc.
    * I don't want a conversion with "bouncy" aesthetics -- where we compel half the extensions to reorganize and then (8 months later) compel them to reorganize again.

This middle-ground allows incremental progress (fix false-matches; improve performance; no reorganization) and avoids rushing the aesthetic questions. (When we're ready to work on ___One Good Layout___, then we can give that proper attention.)

totten committed:
Mixins - Add mgd-php@2. Use tighter search to ignore nested submodules.

totten commented:
I think the test-failure arises because the test-apparatus tries to do some combined tests. (Ex: "What happens if you enable both `mgd-php@1` and `mgd-php@2` -- and test them at the same time?") That's a problem in test-apparatus. I need to look at how to tune it.

colemanw commented:
> @totten Originally, when drafting this PR, I started with that last notion -- to search only `./managed/`. But then I reviewed universe to gauge actual impacts.

That was my impulse too. Good thing you searched `universe`. I'm surprised so many people are *not* using the `managed/` directory. I wonder why?
Part of me still wants to push for strictness and force everyone to use `managed/` but that sounds like a big lift given the variety out in the wild. This compromise seems reasonable.

totten committed:
mixer - Multiple updates

1. Add the option '--exclude=LIST'. This allows us to run a test with
   (say) all mixin EXCEPT foo+bar.

2. Add the option '--copy'. This is actually no different from prior
   behavior. But it makes it easier to read/compare calls. (You
   can distinguish --bare and --copy.)

3. Convert 'LifecyleTest.php' to 'LifecycleTest.template.php'
   (with dataProvider to indicate specific mixins).  This means that you can
   still edit the test-file with syntax-highlighting, but the
   result-sets don't overlap, so Jenkins should do a better
   job of presenting the results.

totten committed:
mixin - Split older/newer mixins into different batches

Situation:

* We have mgd-php@1 and mgd-php@2.
* The tests for mgd-php@2 specifically assert that some -old- behaviors are -not- running.
* In `test-all`, it does isolated tests and combined tests. However, you must not
  combine mgd-php@1 and mgd-php@2 into the same test. This will raise a false-negative for mgd-php@2.

This updates `test-all`. It still does isolated tests, and it still does combined tests.

But the combined tests are split up differnetly, e.g.

* `mixin-combine-bare-new.xml` combines the latest mixins (incl mgd-php@2 but not mgd-php@1).
* `mixin-combine-bare-old.xml` combines the older mixins (incl mgd-php@1 but not mgd-php@2).
